osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Performance;
+using osu.Framework.Graphics.Pooling;
+using osu.Game.Rulesets.Catch.UI;
+using osu.Game.Utils;
+using osuTK;
+
+namespace osu.Game.Rulesets.Catch.Objects.Drawables
+{
+    public class CaughtObjectContainer : CompositeDrawable
+    {
+        public Container<DrawableCaughtObject> StackedObjectContainer { get; }
+
+        private readonly Container<DrawableCaughtObject> droppedObjectTarget;
+
+        private readonly LifetimeEntryManager lifetimeManager = new LifetimeEntryManager();
+
+        private readonly Dictionary<LifetimeEntry, DrawableCaughtObject> drawableMap = new Dictionary<LifetimeEntry, DrawableCaughtObject>();
+
+        private readonly HashSet<CaughtObjectEntry> aliveStackedObjects = new HashSet<CaughtObjectEntry>();
+        private readonly Dictionary<CaughtObjectEntry, CaughtObjectEntry> dropEntryMap = new Dictionary<CaughtObjectEntry, CaughtObjectEntry>();
+
+        private readonly DrawablePool<DrawableCaughtFruit> caughtFruitPool;
+        private readonly DrawablePool<DrawableCaughtBanana> caughtBananaPool;
+        private readonly DrawablePool<DrawableCaughtDroplet> caughtDropletPool;
+
+        /// <summary>
+        /// The randomness used to compute position in stack.
+        /// It is incremented for each <see cref="Add"/> call and decremented for each <see cref="Remove"/> call to make a replay consistent.
+        /// </summary>
+        private int randomSeed = 1;
+
+        public CaughtObjectContainer(Container<DrawableCaughtObject> droppedObjectTarget)
+        {
+            this.droppedObjectTarget = droppedObjectTarget;
+
+            InternalChildren = new Drawable[]
+            {
+                caughtFruitPool = new DrawablePool<DrawableCaughtFruit>(50),
+                caughtBananaPool = new DrawablePool<DrawableCaughtBanana>(100),
+                // less capacity is needed compared to fruit because droplet is not stacked
+                caughtDropletPool = new DrawablePool<DrawableCaughtDroplet>(25),
+                StackedObjectContainer = new Container<DrawableCaughtObject>()
+            };
+
+            lifetimeManager.EntryBecameAlive += entryBecameAlive;
+            lifetimeManager.EntryBecameDead += entryBecameDead;
+        }
+
+        public void Add(CaughtObjectEntry entry)
+        {
+            randomSeed++;
+
+            lifetimeManager.AddEntry(entry);
+
+            if (entry.State != CaughtObjectState.Stacked) return;
+
+            // `DropStackedObjects` may be called before lifetime update.
+            if (entry.LifetimeStart <= Time.Current)
+                aliveStackedObjects.Add(entry);
```

Comment:
This seems slightly suspicious and potentially buggy to me. Can whatever this is fixing be fixed in any other way, like (wild guess) scheduling stacked object drop after children, for instance?

Reply:
Multiple objects can be caught in one frame. `aliveStackObjects` affect stack positioning and we cannot just schedule the explosion to remove this code.
To update `aliveStackObjects` via `entryBecameAlive`, we need to call `lifetimeManager.Update` but this results in quadratic time complexity and is awkward (`LifetimeEntryManager` delays new entry to become alive for reasons).
My current stance is this way is the best.